File: PhysicsSimulator/src/shared/utils/Result.ts
export type Success<T> = {
    isSuccess: true;
    value: T;
};

export type Failure<E> = {
    isSuccess: false;
    error: E;
};

export type FailureResponse<E, R> = {
    isSuccess: false;
    error: E;
    value: R;
};

export type Result<T, E> = Success<T> | Failure<E>;

export type ResultResponse<T, E, R> = Success<T> | FailureResponse<E, R>;

export function success<T>(value: T): Success<T> {
    return { isSuccess: true, value };
}

export function failure<E>(error: E): Failure<E> {
    return { isSuccess: false, error };
}

export function failureResponse<E, R>(error: E, value: R): FailureResponse<E, R> {
    return { isSuccess: false, error, value };
}

File: PhysicsSimulationWebApi/API/Shared/Result.cs
namespace PhysicsSimulationWebApi.Shared;

public class Result
{
    public bool IsSuccess { get; }
    public bool IsFailure => !IsSuccess;
    public Error Error { get; }
    
    protected Result(bool isSuccess, Error error)
    {
        if (isSuccess && error != Error.None || !isSuccess && error == Error.None)
        {
            throw new ArgumentException("Invalid error", nameof(error));
        }

        IsSuccess = isSuccess;
        Error = error;
    }

    public static Result Success() => new(true, Error.None);
    public static Result Failure(Error error) => new(false, error);
}

public class Result<T> : Result
{
    public T Value { get; }

    private Result(bool isSuccess, T value, Error error)
        : base(isSuccess, error)
    {
        Value = value;
    }

    public static Result<T> Success(T value) => new(true, value, Error.None);

    public new static Result<T> Failure(Error error) => new(false, default(T), error);
    public new static Result<T> Failure(Error error, T value) => new(false, value, error);
}